Orm/Xtensive.Orm.Tests/Linq/ContainsOverKeysCollectionTest.cs

```diff
@@ -65,139 +98,593 @@ protected override void PopulateData()
+            .Where(e => new {
+              Key = e.Key,
+              EntityKey = e.Entity.Key,
+              DoubleEntityKey = e.Entity.Entity.Key
+           }.DoubleEntityKey.In(keyArray)).ToList());
+
+        Assert.Throws<QueryTranslationException>(() =>
+          result = session.Query.All<MyEntity3>()
+            .Where(e => new {
+              Key = e.Key,
+               EntityKey = e.Entity.Key,
+              DoubleEntityKey = e.Entity.Entity.Key
+           }.DoubleEntityKey.In(keyArrayAsEnumerable)).ToList());
+
+        Assert.Throws<QueryTranslationException>(() =>
+          result = session.Query.All<MyEntity3>()
+            .Where(e => new {
+              Key = e.Key,
+              EntityKey = e.Entity.Key,
+              DoubleEntityKey = e.Entity.Entity.Key
+            }.DoubleEntityKey.In(keyArrayAsCollection)).ToList());
+      }
+    }
+
+    private void PopulateEntities()
+    {
+      new MyEntity3 {
+        Entity = new MyEntity2 {
+          Entity = new MyEntity { Field = "ABC" }
+        }
+      };
+      new MyEntity3 {
+        Entity = new MyEntity2 {
+          Entity = new MyEntity { Field = "DEF" }
+        }
+      };
+      new MyEntity3 {
+        Entity = new MyEntity2 {
+          Entity = new MyEntity { Field = "GHI" }
+        }
+      };
+      new MyEntity3 {
+        Entity = new MyEntity2 {
+          Entity = new MyEntity { Field = "KLM" }
+        }
+      };
+      new MyEntity3 {
+        Entity = new MyEntity2 {
+          Entity = new MyEntity { Field = "NOP" }
+        }
+      };
+      new MyEntity3 {
+        Entity = new MyEntity2 {
+          Entity = new MyEntity { Field = "QRS" }
+        }
+      };
+
+      new ComplexKeyEntity(1, 1) { Name = "1 1" };
+      new ComplexKeyEntity(2, 2) { Name = "2 2" };
+      new ComplexKeyEntity(3, 3) { Name = "3 3" };
+      new ComplexKeyEntity(4, 4) { Name = "4 4" };
+      new ComplexKeyEntity(5, 5) { Name = "5 5" };
+      new ComplexKeyEntity(6, 6) { Name = "6 6" };
+      new ComplexKeyEntity(7, 7) { Name = "7 7" };
+      new ComplexKeyEntity(8, 8) { Name = "8 8" };
+    }
+
+    private void PopulateKeysCollections(Session session)
+    {
+      keyList = session.Query.All<MyEntity>().AsEnumerable().Take(3).Select(e => e.Key).ToList();
+      keyListAsEnumerable = new List<Key>(keyList);
+      keyListAsCollection = new List<Key>(keyList);
+      keyListAsInteface = new List<Key>(keyList);
+
+      keyArray = session.Query.All<MyEntity>().AsEnumerable().Take(3).Select(e => e.Key).ToArray();
+      keyArrayAsEnumerable = keyArray.AsEnumerable().ToArray();
+      keyArrayAsCollection = keyArray.AsEnumerable().ToArray();
+
+      complexKeyList = session.Query.All<ComplexKeyEntity>().Take(5).Select(e => e.Key).ToList();
+      complexKeyListAsEnumerable = new List<Key>(complexKeyList);
+      complexKeyListAsCollection = new List<Key>(complexKeyList);
+      complexKeyListAsInteface = new List<Key>(complexKeyList);

```

Comment:
I think it is not good to use `List` or array for all cases. If you are testing an interface like `ICollection` or `IReadOnlyCollection` you implementation must be a custom class implementing only the interface you are going to test. This custom class may be a lightweight wrapper over a `List` or an `Array` but it must be completely unique.
In this approach test can guarantee framework doesn't cast this collection to one of those well known types.
Otherwise if the intention was to use casting to `Array` type for example this must be explicitly mentioned in test code.

Reply:
Initially, the test checked how translator detects item type of a given local collection and included only IEnumerable<T> test collections, while working on another problem we've found that it's not possible to use an array without casting it to some generic interface it implements. We decided it is wierd and native usage of array should work too so test was extended with additional representations of native generic type (in this case ```List<T>```) and array. Custom non-generic class implementing one of collection interfaces was forgotten though. I will add it to the test and adjist Item type detection.

To summarize, to me we need following test collections:
- an array used natively (check for array must be separate because Type.IsArray is faster that check for interface implementation);
- a custom generic class implementing ```IEnumerable<T>```, since majority of collections implement it eventually (directly or throug other interfaces from System.Collections.Generic namespace);
- custom non-generic class implementing ```IEnumerable<T>``` to make sure it works fine too.